ad5770r: Initial commit of AD5770R driver

mirceacaprioru opened the pull request:
The AD5770R is a 6-channel, 14-bit resolution, low noise, programmable
current output digital-to-analog converter (DAC) for photonics control
applications. The chip contains five 14-bit resolution sourcing DAC
channels and one 14-bit resolution current sourcing/sinking DAC channel.

The driver enables the user to set the output of any of the five DAC
channels with hardware LDAC control or software LDAC.

Signed-off-by: Mircea Caprioru <mircea.caprioru@analog.com>

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+			     uint8_t reg_addr,
+			     uint8_t *reg_data)
+{
+	uint8_t buf[3];
```
nitpick: you can make this `buf[2]` ; since you are reading/writing only 2 bytes

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	buf[1] = 0x00;
+
+	ret = spi_write_and_read(dev->spi_desc, buf, 2);
+	*reg_data = buf[1];
```
since this function can be called outside of this file [as an API or other] maybe add a checks for `dev` and `reg_data`
similarly, this can be done for other similar functions;
usually functions return `-EINVAL` when nulls are passed as args

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	/* Check if multibyte enabled */
+
+	buf[0] = AD5770R_REG_READ(reg_addr);
+	memset(&buf[1], 0x00, count);
```
in this case, maybe make sure `count < sizeof(buf) - 1`

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	ret = spi_write_and_read(dev->spi_desc, buf, count + 1);
+
+	for (i = 0; i < count; i++)
+		reg_data[i] = buf[i+1];
```
you can use `memcpy()` here

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	buf[0] = AD5770R_REG_WRITE(reg_addr);
+	buf[1] = reg_data & 0xFF;
+
+	ret = spi_write_and_read(dev->spi_desc, buf, 2);
```
nitpick: more of a style-comment; you can do `return spi_write_and_read(dev->spi_desc, buf, 2);` directly here

commodo replied:
also, a nitpick [feel free to disregard]: you could pass `sizeof(buf)` instead of `2`; since many other drivers already pass constant values, I'm not feeling strongly about this ; so, i'll leave this up to you

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+				       uint16_t count)
+{
+	int32_t ret;
+	uint8_t data[1024];
```
this also applies to other similar cases; but 1024 bytes on the call-stack can be a lot [depending on the user/application];
if we consider applications [like a few in the automotive], their use of the stack is limited via coding standard [or at least it used to be]; or if we consider some deployments that include micro-controllers ;

maybe using malloc() here would be slightly better;
but I'm open for discussion here;

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	data[0] = AD5770R_REG_WRITE(reg_addr);
+	memcpy(&data[1], reg_data, count);
+
+	ret = spi_write_and_read(dev->spi_desc, data, count + 1);
```
you could do `return spi_write_and_read(dev->spi_desc, data, count + 1);`   here directly;

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	reg_data |= data;
+	ret |= ad5770r_spi_reg_write(dev, reg_addr, reg_data);
+
+	return ret;
```
`return ad5770r_spi_reg_write(dev, reg_addr, reg_data);`

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_device_spi(struct ad5770r_dev *dev,
+			       struct device_spi_settings spi_settings)
```
pass this via const pointer;
```
int32_t ad5770r_set_device_spi(struct ad5770r_dev *dev,
			       const struct device_spi_settings *spi_settings)
```

the way this is done now, copies the entire `struct device_spi_settings` on the call-stack

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	dev->dev_spi_settings.stream_mode_length =
+		spi_settings.stream_mode_length;
+
+	return ret;
```
if you don't plan to add anything later here, this could be `return 0` ;

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+				    AD5770R_HW_LDAC_MASK_CH(mask_hw_ldac[5], AD5770R_CH5));
+
+	if (ret)
+		return -1;
```
maybe do,  
```
if (ret)
    return ret;
```
in all places;
that way we pass the error code up to the caller; so it doesn't get lost

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+ */
+int32_t ad5770r_set_output_mode(struct ad5770r_dev *dev,
+				struct output_range output_mode,
+				enum ad5770r_channels channel)
```
same for this struct; pass it via pointer;

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+
+	ret = ad5770r_spi_reg_write_multiple(dev,
+					     AD5770R_DAC_PAGE_MASK_MSB,
+					     data, 2);
```
maybe it would be a good idea to do a `if (ret) return ret;`  here

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_init(struct ad5770r_dev **device,
+		     struct ad5770r_init_param init_param)
```
pass this via pointer

lclausen-adi commented on drivers/ad5770r/ad5770r.h:
```diff
@@ -0,0 +1,377 @@
+	bool			addr_ascension;
+	bool			single_instruction; // for multibyte read/write
+	uint8_t		stream_mode_length;
+};
```
What about those structs above, should they be prefixed to avoid namespace issues?

commodo commented on drivers/ad5770r/ad5770r.h:
```diff
@@ -0,0 +1,377 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define BIT(x)					(1UL << (x))
```
it's a good idea to have this guarded via #ifndef
```
#ifndef BIT
#define BIT(x)
#endif
```
to prevent this colliding with other similar definitions

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	else
+		printf("ad5770r initialization error (%d)\n", ret);
+
+	mdelay(1000);
```
maybe remove this delay; some users may not want it; or would like to add their own delay

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_mask_channel(struct ad5770r_dev *dev,
+				 bool *mask_channel_sel)
```
i'm seeing a few cases where a predefined array of `bool` ( with 6 elements or more ) is passed ;
it's not a good idea to leave it like this for others to use, as it can be ambiguous for users about the number of elements in the array they should use;

there are 2 ways to do this:
1) document this [in the header or other docs ]
2) enforce a fixed typedef/struct;

[An example] You could have, 
```
struct ad5770r_channel_switches {
     bool en0, en1, en2, en3, en4, en5;
}
```
then
```
int32_t ad5770r_set_mask_channel(struct ad5770r_dev *dev,
				 struct ad5770r_channel_switches *ch)

ret = ad5770r_spi_reg_write(dev,
				    AD5770R_CH_SELECT,
				    AD5770R_CH_SELECT_SEL_CH(ch->en0, AD5770R_CH0) |
				    AD5770R_CH_SELECT_SEL_CH(ch->en1, AD5770R_CH1) |
				    AD5770R_CH_SELECT_SEL_CH(ch->en2, AD5770R_CH2) |
				    AD5770R_CH_SELECT_SEL_CH(ch->en3, AD5770R_CH3) |
				    AD5770R_CH_SELECT_SEL_CH(ch->en4, AD5770R_CH4) |
				    AD5770R_CH_SELECT_SEL_CH(ch->en5, AD5770R_CH5));
```

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+	if (ret)
+		return -1;
+
+	dev->mask_channel_sel = mask_channel_sel;
```
also, passing references like this here, should be documented, or you should keep a copy of the data in the driver;
if the user passes a value that is on the stack, the `mask_channel_sel` pointer will point to something invalid at some point in time;

lclausen-adi replied:
We just shouldn't do it. Taking references to other objects and keeping the reference without any object lifetime management is asking for trouble.

Besides the driver never uses this, so maybe just drop it.

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,650 @@
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_monitor_setup(struct ad5770r_dev *dev,
+				  struct monitor_setup mon_setup)
```
pass this via pointer; this also allocs `sizeof(struct monitor_setup)` on the call-stack

commodo commented on drivers/ad5770r/ad5770r.h:
```diff
@@ -0,0 +1,377 @@
+int32_t ad5770r_init(struct ad5770r_dev **device,
+		     struct ad5770r_init_param init_param);
+
+#endif // AD5770R_H_0
```
nitpick: this looks like a typo;  `AD5770R_H_0` -> `AD5770R_H_`; 
since this line may change, maybe use comment style `/* AD5770R_H_ */`

mirceacaprioru commented:
Changelog V1 -> V2

- replaced return -1 with return ret
- added prefix to structures
- replaced bool arrays with struct ad5770r_channel_switches
- device structure no longer takes references to objects 
- replaced malloc with calloc
- 1024 arrays are now represented by a pointer that has memory allocated depending on count value. 
- removed delay at the end of init function
- added checks for dev and reg_data NULL pointer case

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	int32_t ret;
+
+	if (!dev | !reg_data)
+		return FAILURE;
```
why not `-EINVAL` ?

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	buf[0] = AD5770R_REG_READ(reg_addr);
+	buf[1] = 0x00;
+
+	ret = spi_write_and_read(dev->spi_desc, buf, sizeof(buf));
```
do error checking here ?
```
if (ret)
    return ret;
```

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	buf[0] = AD5770R_REG_READ(reg_addr);
+	memset(&buf[1], 0x00, count);
+
+	ret = spi_write_and_read(dev->spi_desc, buf, count + 1);
```
error checking here ?

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	buf = (uint8_t*)calloc(count + 1, sizeof(uint8_t));
+
+	buf[0] = AD5770R_REG_READ(reg_addr);
+	memset(&buf[1], 0x00, count);
```
`memset(0)` is redundant if you're using `calloc()`

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+
+	ret = spi_write_and_read(dev->spi_desc, buf, count + 1);
+
+	memcpy(reg_data, buf + 1, count);
```
you need to `free(buf)` here, and on all exit paths if changing stuff

lclausen-adi replied:
Long term it might be good to add a spi\_write\_**then**\_read() to our set of library functions. The function would take two buffers and two counts, one for write and one for read.

This seems to be a common pattern and would make the whole alloc/memcpy/free dance unnecessary.

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	data = (uint8_t*)calloc(count + 1, sizeof(uint8_t));
+
+	data[0] = AD5770R_REG_WRITE(reg_addr);
+	memcpy(&data[1], reg_data, count);
```
`free(data)` before exiting

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	if (ret)
+		return ret;
+
+	dev->channel_config = *channel_config;
```
this still passes a reference to an external object, which may have limited life-time

commodo replied:
wait; my bad
it de-references the object;

commodo replied:
it would be a good idea to do `memcpy(&dev->channel_config, channel_config, sizeof(dev->channel_config))`
this is more certain to work well on most C compilers;

lclausen-adi replied:
No, keep it the way it is. This should work with all compilers either way.

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+			    const struct ad5770r_channel_switches *mask_hw_ldac)
+{
+	int32_t ret;
+
```
you could to null-checks here:
```
if (!dev || !mask_hw_ldac)
    return -EINVAL;
```
and in similar places

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+
+	dev = (struct ad5770r_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return FAILURE;
```
you could use `-ENOMEM` here

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	uint8_t product_id_l, product_id_h;
+	int32_t ret;
+	enum ad5770r_channels i;
+
```
you could check if `device` is NULL here;

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+
+	/* Device settings */
+	ret = ad5770r_set_device_spi(dev,
+				     &init_param->dev_spi_settings);
```
there is no case to handle  `init_param` being null;
there are 2 ways to do it:
1) if `init_param` is NULL, return -EINVAL
2) if `init_param` is NULL pass a default static `init_param` with default values for this device

your preference which one to choose

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,658 @@
+	ret |= ad5770r_set_hw_ladc(dev,
+				   &init_param->mask_hw_ldac);
+
+	for( i = AD5770R_CH0; i <=  AD5770R_CH5; i++) {
```
hmm; maybe also run a style-check on this;

mirceacaprioru commented:
Changelog V2 -> V3

- added checks for dev and other necessary structures
- added free(data) before exiting function
- remove memset(0) after calloc

commodo commented on drivers/ad5770r/ad5770r.c:
```diff
@@ -0,0 +1,718 @@
+			   uint8_t *status)
+{
+
+	if (!dev)
```
nitpick: extra line before this check;
also, `ad5770r_spi_reg_read()` handles NULL, so you could remove this check
